### ISSUE TITLE
Possible deadlock in sendStream and framerI

### ISSUE THREAD

**Issue by VictoriaRaymond** [lucas-clemente/quic-go#1658]
I got the following callstacks showing deadlocks in sendStream and framerI, although I can't prove they are the same objects across goroutines.

The symptom is `framerI` holds its own lock at `framerI.AppendStreamFrames`, and then tries to lock `sendStream` at `sendStream.popStreamFrameImpl`, while `sendStream.Close` already holds the lock, and then tries to lock `framerI` at `AddActiveStream`.

```
13 @ 0x102d57b 0x102d623 0x103e06c 0x103dd9d 0x10636cf 0x1578c81 0x158a8c4 0x158aadd 0x1580e52 0x158af13 0x15959c0 0x14fd99b 0x137a507 0x135692b 0x105b461
#	0x103dd9c	sync.runtime_SemacquireMutex+0x3c								go/src/runtime/sema.go:71
#	0x10636ce	sync.(*Mutex).Lock+0xfe										go/src/sync/mutex.go:134
#	0x1578c80	github.com/lucas-clemente/quic-go.(*framerI).AddActiveStream+0x30		github.com/lucas-clemente/quic-go/framer.go:68
#	0x158a8c3	github.com/lucas-clemente/quic-go.(*session).onHasStreamData+0x43		github.com/lucas-clemente/quic-go/session.go:1118
#	0x158aadc	github.com/lucas-clemente/quic-go.(*uniStreamSender).onHasStreamData+0x3c	github.com/lucas-clemente/quic-go/stream.go:35
#	0x1580e51	github.com/lucas-clemente/quic-go.(*sendStream).Close+0x121		github.com/lucas-clemente/quic-go/send_stream.go:225
#	0x158af12	github.com/lucas-clemente/quic-go.(*stream).Close+0x32			github.com/lucas-clemente/quic-go/stream.go:133
#	0x15959bf	v2ray.com/core/transport/internet/quic.(*interConn).Close+0x4f					v2ray.com/core/transport/internet/quic/conn.go:186
```

```
1 @ 0x102d57b 0x102d623 0x103e06c 0x103dd9d 0x10636cf 0x1580834 0x1580789 0x1578f6b 0x157d0aa 0x157c4f3 0x1589653 0x1588a3a 0x1585130 0x1590ae1 0x105b461
#	0x103dd9c	sync.runtime_SemacquireMutex+0x3c									go/src/runtime/sema.go:71
#	0x10636ce	sync.(*Mutex).Lock+0xfe											go/src/sync/mutex.go:134
#	0x1580833	github.com/lucas-clemente/quic-go.(*sendStream).popStreamFrameImpl+0x43		github.com/lucas-clemente/quic-go/send_stream.go:146
#	0x1580788	github.com/lucas-clemente/quic-go.(*sendStream).popStreamFrame+0x38		github.com/lucas-clemente/quic-go/send_stream.go:138
#	0x1578f6a	github.com/lucas-clemente/quic-go.(*framerI).AppendStreamFrames+0x1da		github.com/lucas-clemente/quic-go/framer.go:95
#	0x157d0a9	github.com/lucas-clemente/quic-go.(*packetPacker).composeNextPacket+0x109		github.com/lucas-clemente/quic-go/packet_packer.go:390
#	0x157c4f2	github.com/lucas-clemente/quic-go.(*packetPacker).PackPacket+0x142		github.com/lucas-clemente/quic-go/packet_packer.go:291
#	0x1589652	github.com/lucas-clemente/quic-go.(*session).sendPacket+0x82			github.com/lucas-clemente/quic-go/session.go:961
#	0x1588a39	github.com/lucas-clemente/quic-go.(*session).sendPackets+0x139			github.com/lucas-clemente/quic-go/session.go:852
#	0x158512f	github.com/lucas-clemente/quic-go.(*session).run+0x45f				github.com/lucas-clemente/quic-go/session.go:417
#	0x1590ae0	github.com/lucas-clemente/quic-go.(*client).establishSecureConnection.func1+0x40	github.com/lucas-clemente/quic-go/client.go:274
```

**Comment by marten-seemann** [comment]
Hello @VictoriaRaymond, thanks for the detailed bug report and sorry for the late reply. I'm currently on vacation, and I will have a look at it as soon as I come back.

**Comment by marten-seemann** [comment]
Hi @VictoriaRaymond, sorry for the late reply. I just ran into the same issue when writing an integration test for my fix for #1662. I still don't fully understand what exactly is going wrong here, but I'm working on that. Thanks again for your bug report!

**Comment by Xyncgas** [comment]
VictoriaRaymond, creator of v2ray, the proxy program that allows chinese to bypass the internet censorship by their goverment... is she dead? is she captured by the communist? #DeveloperLivesMatter
